packages/core/src/internal/sagaError.js

```diff
@@ -0,0 +1,77 @@
+// there can be only a single saga error created at any given moment
+// so this module acts as a singleton for bookkeeping it
+import { getLocation, flatMap } from './utils'
+
+function formatLocation(fileName, lineNumber) {
+  return `${fileName}?${lineNumber}`
+}
+
+function effectLocationAsString(effect) {
+  const location = getLocation(effect)
+  if (location) {
+    const { code, fileName, lineNumber } = location
+    const source = `${code}  ${formatLocation(fileName, lineNumber)}`
+    return source
+  }
+  return ''
+}
+
+function sagaLocationAsString(sagaMeta) {
+  const { name, location } = sagaMeta
+  if (location) {
+    return `${name}  ${formatLocation(location.fileName, location.lineNumber)}`
+  }
+  return name
+}
+
+function cancelledTasksAsString(sagaStack) {
+  const cancelledTasks = flatMap(i => i.cancelledTasks, sagaStack)
+  if (!cancelledTasks.length) {
+    return ''
+  }
+  return ['Tasks cancelled due to error:', ...cancelledTasks].join('\n')
+}
+
+const UNIQUE = {}
+let crashedEffect = UNIQUE
+const sagaStack = []
+
+export const addSagaFrame = frame => {
+  sagaStack.push(frame)
+}
+
+export const clear = () => {
+  crashedEffect = UNIQUE
+  sagaStack.length = 0
+}
+
+export const setCrashedEffect = effect => {
+  // track only first effect (the one which has caused the crash)
+  if (crashedEffect !== UNIQUE) {
```

Comment:
after a thought - falsy values cannot crash saga, so the whole `UNIQUE` thing can be removed, and a simple truthy check should suffice